### PR TITLE
Add libsodium resolver

### PR DESCRIPTION
**Pull request by popzxc**
This PR adds a new resolver based on `libsodium-sys`/`sodiumoxide`. 

Unfortunately, current state of `sodiumoxide` lacks some features (e.g. `crypto_sign_ed25519_pk_to_curve25519` function), so code looks a bit raw.
Also the last release does not have `Clone` trait implemented for `State` structure in `Sha256`, so we have to use concrete revision.

However, I'm working on implementing necessary changes to `sodiumoxide` and as soon as they're added I'll update this implementation also.

### CODE REVIEW

**Comment by popzxc** [comment]
Hmm... I guess there are some problems with `appveyor` run and I have to update script to install `libsodium-sys`. I'll add an `work in progress` label until this problem is resolved.

**Comment by popzxc** [comment]
Well... I've got all the things working (and disabled libsodium resolver testing on appveyor until they'll update `build.rs`).
But now one test fails for some reason. I'm investigating this.
Seems like there is some issue with `ChaChaPoly`.

**Comment by mcginty** [comment]
Yeah, that's very interesting and I while there are tests in the specific resolvers to check their crypto implementations against known results, think to truly solve this we can't have the initiator and responder using the same Noise implementation. When you're developing are you testing against the external vectors? Those should have caught the issue.

Another great addition to snow would be some kind of external test scaffolding similar to what WireGuard does, so we can test cross-implementation handshakes against the appropriate latest versions of other implementations.

File: src/resolvers/libsodium.rs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+        32
+    }
+
+    fn priv_len(&self) -> usize {
+        32
+    }
+
+    fn set(&mut self, privkey: &[u8]) {
+        self.privkey = sodium_curve25519::Scalar::from_slice(privkey)
+            .expect("Can't construct private key for Dh25519");
```
Comment (mcginty) [inline src/resolvers/libsodium.rs]: when exactly would this cause a panic?

File: tests/general.rs [hunk 2]
```diff
@@ -371,7 +371,7 @@ fn test_transport_message_exceeds_max_len() {
     let mut noise = NoiseBuilder::new(params).remote_public_key(&[0u8; 32]).build_initiator().unwrap();
 
     let mut buffer_out = [0u8; 65535*2];
-    noise.write_message(&[0u8; 0], &mut buffer_out).unwrap();
+    let _res = noise.write_message(&[0u8; 0], &mut buffer_out);
```
Comment (mcginty) [inline tests/general.rs]: what's the reason for this change to ignore the result of `write_message`?

File: Cargo.toml [hunk 3]
```diff
@@ -43,6 +45,8 @@ rust-crypto = { version = "^0.2", optional = true }
 hacl-star = { version = "=0.0.11", optional = true }
 rand = { version = "0.5.0-pre.2", optional = true }
 ring = { version = "0.13.0-alpha5", optional = true }
+#sodiumoxide = { version = "^0.1.0", optional = true } # We have to use concrete revision until version bump.
+sodiumoxide = { git = "https://github.com/sodiumoxide/sodiumoxide", rev = "d01bbdbd91313453ae244d56973545d9b695af20", optional = true }
```
Comment (mcginty) [inline Cargo.toml]: Like you mentioned earlier, we're going to have to wait to merge this until there's a released version of `sodiumoxide` since I can't in turn release `snow` if it depends on any git-sourced crates. Hopefully it doesn't take them too long :).

File: src/resolvers/libsodium.rs (hunk 1 above)
Comment (popzxc) [reply src/resolvers/libsodium.rs]: When slice of incorrect length is passed, for example.

File: tests/general.rs (hunk 2 above)
Comment (popzxc) [reply tests/general.rs]: Because some implementations of `dh` may fail, and some may not. We can't for now check if result is `Err` or `Ok`.

File: Cargo.toml (hunk 3 above)
Comment (popzxc) [reply Cargo.toml]: Yes, sure. This PR is not "to-be-merged-asap", but to discuss its state and maybe polish it.

**Comment by popzxc** [comment]
Regarding test vectors - yes, I used them, and still got only 1 test failing.

**Comment by mcginty** [comment]
@popzxc I just ran the tests and the vectors fail hard with the wrong implementation, so things are working as expected. What happened for you was that the vector-tests never ran since an earlier test suite already failed and rust doesn't run all suites regardless of the past results.

So, if you run `cargo test --features libsodium-accelerated,vector-tests`, you'd just see the one known-result test fail, but if you ran `cargo test --features libsodium-accelerated,vector-tests vector`, you'll see all the related vectors fail :).

File: tests/general.rs (hunk 2 above)
Comment (mcginty) [reply tests/general.rs]: I'm a bit confused, wouldn't we want the test to fail if the dh fails?

File: src/resolvers/libsodium.rs (hunk 1 above)
Comment (mcginty) [reply src/resolvers/libsodium.rs]: Cool thanks. Seems reasonable.

**Comment by popzxc** [comment]
Ah, I think I got this. OK then, my bad :)

File: tests/general.rs (hunk 2 above)
Comment (popzxc) [reply tests/general.rs]: Nope, since it's an incorrect message (filled with only zeros), it's okay for dh to fail (as far as I know).

File: src/resolvers/libsodium.rs [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+use self::sodiumoxide::crypto::aead::chacha20poly1305_ietf as sodium_chacha20poly1305;
+use self::sodiumoxide::crypto::hash::sha256 as sodium_sha256;
+use self::sodiumoxide::crypto::scalarmult::curve25519 as sodium_curve25519;
+
+#[derive(Default)]
+pub struct SodiumResolver;
+
+impl CryptoResolver for SodiumResolver {
+    fn resolve_rng(&self) -> Option<Box<Random + Send>> {
+        None
```
Comment (mcginty) [inline src/resolvers/libsodium.rs]: Any reason we don't want the included RNG in sodiumoxide?

File: src/resolvers/libsodium.rs [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+// Elliptic curve 25519.
+pub struct SodiumDh25519 {
+    privkey: sodium_curve25519::Scalar,
+    pubkey: sodium_curve25519::GroupElement,
+}
+
+impl SodiumDh25519 {
+    fn convert_to_private_key(key: &mut [u8; 32]) {
+        key[0] &= 248;
```
Comment (mcginty) [inline src/resolvers/libsodium.rs]: this function is only used once, can just inline it in `generate()`?

File: src/resolvers/libsodium.rs [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    }
+
+    fn privkey(&self) -> &[u8] {
+        &self.privkey[0..32]
+    }
+
+    fn dh(&self, pubkey: &[u8], out: &mut [u8]) -> Result<(), ()> {
+        let pubkey = sodium_curve25519::GroupElement::from_slice(&pubkey[0..32])
+            .expect("Can't construct public key for Dh25519");
+        let result = sodium_curve25519::scalarmult(&self.privkey, &pubkey);
```
Comment (mcginty) [inline src/resolvers/libsodium.rs]: I think I'd prefer 

```rust
let result = sodium_curve25519::scalarmult(&self.privkey, &pubkey)
    .map_err(|_| ())?;
out[..32].copy_from_slice(&buf[0..32]);
Ok(())
```

instead of the match statement.

File: src/resolvers/libsodium.rs [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+        ciphertext: &[u8],
+        out: &mut [u8],
+    ) -> Result<usize, ()> {
+        let mut nonce_bytes = [0u8; 12];
+        LittleEndian::write_u64(&mut nonce_bytes[4..], nonce);
+        let nonce = sodium_chacha20poly1305::Nonce(nonce_bytes);
+
+        let result = sodium_chacha20poly1305::open(ciphertext, Some(authtext), &nonce, &self.key);
+
+        match result {
```
Comment (mcginty) [inline src/resolvers/libsodium.rs]: same comment as above with the `map_err(|_| ())` preference

File: src/resolvers/libsodium.rs (hunk 4 above)
Comment (popzxc) [reply src/resolvers/libsodium.rs]: Hm... I missed somehow that `sodiumoxide` provides a public interface for that. I'll add it soon then.

File: src/resolvers/libsodium.rs (hunk 5 above)
Comment (popzxc) [reply src/resolvers/libsodium.rs]: IMHO conversion logic should be encapsulated and not mixed up with key generation.
Since it's used only once, it'll be inlined by the compiler, and `convert_*` call describes what's going on better than some binary operations, even with the comment.

`libsodium` contains separate function for such a conversion, and as soon as it's added to the `sodiumoxide` (I'm going to add corresponding PR), this method will be removed.

Comment (mcginty) [reply src/resolvers/libsodium.rs]: Fair argument, will be a moot point (and aligned with other x25519 libraries like dalek) once they add the function in themselves :).

**Comment by mcginty** [comment]
Should the resolver be calling sodiumoxide's `init()` function at some point?

**Comment by popzxc** [comment]
Nope, I don't think so. It's mentioned in `sodiumoxide` that `sodiumoxide::init()` should be called before any other function from `sodiumoxide`. I guess that if user choose `libsodium-resolver`, he either already uses it in its project (and calls this function on already) or know what he's doing (and will call this function at the program start).

So I think a note in README.md and documentation will be enough.

**Comment by mcginty** [comment]
>Nope, I don't think so. It's mentioned in `sodiumoxide` that `sodiumoxide::init()` should be called before any other function from `sodiumoxide`. I guess that if user choose `libsodium-resolver`, he either already uses it in its project (and calls this function on already) or know what he's doing (and will call this function at the program start). ...

@popzxc I think we should not assume people are necessarily using libsodium outside of their use of snow.

Put another way: is there a downside to running `init()` whenever the module is loaded using something like `lazy_static!`?

**Comment by mcginty** [comment]
Closing in favor of #75, an effort to revive this PR. Thanks again for kicking things off, @popzxc.
